tests/integration/test_backup_restore_on_cluster/test_really_many_tables.py

```diff
@@ -0,0 +1,103 @@
+import time
+
+import pytest
+
+from helpers.cluster import ClickHouseCluster
+
+cluster = ClickHouseCluster(__file__)
+
+main_configs = [
+    "configs/cluster.xml",
+    "configs/backups_disk.xml",
+]
+
+user_configs = []
+
+image = "clickhouse/integration-test"
+tag = None
+
+#image = "clickhouse/clickhouse-server"
+#tag = "24.10"
+
+node1 = cluster.add_instance(
+    "node1",
+    image=image,
+    tag=tag,
+    main_configs=main_configs,
+    user_configs=user_configs,
+    external_dirs=["/backups/"],
+    macros={"replica": "node1", "shard": "shard1"},
+    with_zookeeper=True,
+)
+
+node2 = cluster.add_instance(
+    "node2",
+    image=image,
+    tag=tag,
+    main_configs=main_configs,
+    user_configs=user_configs,
+    external_dirs=["/backups/"],
+    macros={"replica": "node2", "shard": "shard1"},
+    with_zookeeper=True,
+)
+
+
+@pytest.fixture(scope="module", autouse=True)
+def start_cluster():
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+@pytest.fixture(autouse=True)
+def drop_after_test():
+    try:
+        yield
+    finally:
+        node1.query("DROP TABLE IF EXISTS tbl ON CLUSTER 'cluster' SYNC")
+
+
+backup_id_counter = 0
+
+
+def new_backup_name():
+    global backup_id_counter
+    backup_id_counter += 1
+    return f"Disk('backups', '{backup_id_counter}')"
+
+
+def test_really_many_tables():
+    num_tables = 500
+
+    node1.query(
+        "CREATE DATABASE mydb ON CLUSTER 'cluster' ENGINE=Replicated('/clickhouse/path/','{shard}','{replica}')"
+    )
+
+    for i in range(1, num_tables):
+        node1.query(
+            f"CREATE TABLE mydb.tbl{i}(x UInt8, y String) ENGINE=ReplicatedMergeTree ORDER BY x"
+        )
+
+    node2.query("SYSTEM SYNC DATABASE REPLICA mydb")
+
+    # Make backup.
+    backup_name = new_backup_name()
+    old_time = time.monotonic()
+    node1.query(f"BACKUP DATABASE mydb ON CLUSTER 'cluster' TO {backup_name}")
+    backup_time = time.monotonic() - old_time
+
+    print(f"Making backup of {num_tables} tables took {backup_time} seconds")
+
+    # Drop table on both nodes.
+    node1.query("DROP DATABASE mydb ON CLUSTER 'cluster' SYNC")
+
+    # Restore from backup.
+    old_time = time.monotonic()
+    node1.query(f"RESTORE DATABASE mydb ON CLUSTER 'cluster' FROM {backup_name}")
+    restore_time = time.monotonic() - old_time
+
+    print(f"Restoring of {num_tables} tables took {restore_time} seconds")
+
+    assert False
```

Comment:
I got the following results locally:

24.8:
```
Making backup of 500 tables took 1.1659625370002686 seconds
Restoring of 500 tables took 14.490106503999414 seconds

Making backup of 500 tables took 1.566683947000456 seconds
Restoring of 500 tables took 14.842752201000621 seconds

Making backup of 500 tables took 1.2162339060005252 seconds
Restoring of 500 tables took 14.443287267000414 seconds

Making backup of 500 tables took 1.266393320998759 seconds
Restoring of 500 tables took 15.144840867000312 seconds
```

24.10:
```
Making backup of 500 tables took 1.215970336999817 seconds
Restoring of 500 tables took 16.54580684299981 seconds

Making backup of 500 tables took 1.7170670929999687 seconds
Restoring of 500 tables took 14.94252839299952 seconds

Making backup of 500 tables took 1.26649032899968 seconds
Restoring of 500 tables took 15.241242956999486 seconds

Making backup of 500 tables took 1.4163817149983515 seconds
Restoring of 500 tables took 12.58738621900011 seconds
```

master with [PR #72427: Create tables in parallel while restoring](https://github.com/ClickHouse/ClickHouse/pull/72427):
```
Making backup of 500 tables took 1.3160250720002296 seconds
Restoring of 500 tables took 12.437472751000314 seconds

Making backup of 500 tables took 1.1657633650002026 seconds
Restoring of 500 tables took 14.190329081000527 seconds

Making backup of 500 tables took 1.166175163999469 seconds
Restoring of 500 tables took 14.240233277999323 seconds

Making backup of 500 tables took 1.2158479330000773 seconds
Restoring of 500 tables took 14.341145684999901 seconds
```

master without [PR #72427: Create tables in parallel while restoring](https://github.com/ClickHouse/ClickHouse/pull/72427):
```
Making backup of 500 tables took 1.165624758001286 seconds
Restoring of 500 tables took 14.690316699001414 seconds

Making backup of 500 tables took 1.3661946120009816 seconds
Restoring of 500 tables took 15.593186721000166 seconds
```